### ISSUE TITLE
Request for `c_dynamic_array` .. and array views?

### ISSUE THREAD

**Issue by BryantLam** [chapel-lang/chapel#16169]
Feature request for something akin to `c_dynamic_array` as a record wrapper around `c_malloc` and/or `c_calloc` + size so users that are managing heap-allocated C arrays from Chapel can easily create and automatically clean up without leaking memory.

One design question could be supplying custom allocation or deallocation methods similar to customizing options on C++ `std::vector`. This would allow users to specially allocate and deallocate for their platform's quirks.

Further generalizing, the allocation and management of the underlying memory region could be separated from the notion of accessing it. Similar to how Chapel has slicing syntax on arrays, a notion of "array views" could be added to allow arbitrary sub-array slicing to (1) access said allocation via a default array view that encompasses the entire memory region or (2) create a sub-array-view slice.

**Comment by mppf** [comment]
I'm interested in knowing to what extent these two existing features meet this request.

First, supposing we had `c_dynamic_array` that did not store a size, how is it different from `c_ptr`? Are you hoping it handles multidimensional arrays?

Second, we have `chpl_external_array` which currently consists of a pointer to elements, number of elements, and a function to call to free the memory. My hope has been that we would expand this to include the equivalent of a numpy array or [CFI_cdesc_t](https://wg5-fortran.org/N1851-N1900/N1869.pdf). These are generalized multidimensional array representations that support strided access as well. By specifying a no-op free function they could represent a slice (provided the programmer does not delete the original array before the slice is used). It's currently possible to construct a Chapel array that aliases the memory of a `chpl_external_array` (it's currently a DefaultRectangular array). If we continue with this direction, would it meet the feature request here?

**Comment by BryantLam** [comment]
> First, supposing we had `c_dynamic_array` that did not store a size, how is it different from `c_ptr`? Are you hoping it handles multidimensional arrays?

That's a good question. It would know how to free itself, but you'd still need to carry around the size as an argument. Maybe that's fine though.

Maybe multidimensional arrays could be handled, but that seems like it could be walking down a much more complicated path given how application-specific it could be; e.g., some hypothetical triangular matrix consisting of a heap-allocated array with pointers to stack-allocated arrays of increasing sizes.

In any case, a `c_dynamic_array` type maybe would need to be exposed to C code as a struct definition with a stable API to access it. That could be a consideration to make it as featureful as needed if the C code doesn't necessarily care about how that array is represented.

That said, I'm also open to the idea that this feature isn't strictly necessary since it could get confusing on when such an array would be automatically freed in Chapel code, which might cause use-after-free problems in Chapel or in C.

> By specifying a no-op free function they could represent a slice (provided the programmer does not delete the original array before the slice is used). It's currently possible to construct a Chapel array that aliases the memory of a `chpl_external_array` (it's currently a DefaultRectangular array). If we continue with this direction, would it meet the feature request here?

Maybe? This feature request was generated mostly from toying around with [this program](https://github.com/BryantLam/chapel-toys/blob/fa43bf51f01bded36ff37e96734a0f4cf29e7213/src/StringSerialize.chpl) where, ideally, I could have the `deserialize` routine automatically deallocate the array at end of scope. It's done manually in the code right now, but I don't have a quick intuition on whether `buffer`'s implicit argument intent would hypothetically be correct for it to be done automatically if it were some `c_dynamic_array`. (My code isn't great though, so there's likely a better way to do it.)

**Comment by bradcray** [comment]
> you'd still need to carry around the size as an argument. Maybe that's fine though.

Alternatively, we could store a (per-locale) map from `c_dynamic_array` pointers to `size` to avoid storing it with the type if we didn't want to do that.

> Maybe multidimensional arrays could be handled

If it's not part of Bryant's vision (as it sounds), I'd like to propose squashing this path of the conversation.  C doesn't really have multidimensional heap-allocated arrays to speak of, so for me, it starts to get away from the notion of what a c_dynamic_array would be.

The potential value I see in this proposal is that since we introduced `c_array` it seems like the number of people who've said "Oh, I have to use a compile-time size" has been quite large; so this could introduce a sibling type that didn't have to have a param size, was allocated on the stack, but otherwise worked quite similarly.  Whereas `c_ptr` is obviously a bit more manual.  To that extent, for questions like "When would it be freed?" I'd think we'd want the same answer as for `c_array`.

**Comment by mppf** [comment]
> In any case, a c_dynamic_array type maybe would need to be exposed to C code as a struct definition with a stable API to access it. That could be a consideration to make it as featureful as needed if the C code doesn't necessarily care about how that array is represented.

That's what `chpl_external_array` is - it's a C struct - that might not have been clear before.

>  To that extent, for questions like "When would it be freed?" I'd think we'd want the same answer as for c_array.

This is the part where the feature request here is different from `c_ptr` for me. A `c_ptr` will never deallocate the buffer when it goes out of scope.

I think it'd be better to have one C struct representing a Chapel array for interoperability than two. (Well... we already have 3 right now, arguably; I'm just saying I don't think we should add another one). So I'd rather, for example, rename `chpl_external_array` to `c_dynamic_array`. (Of course if either of these were actually a C struct and we wanted it to be freed when going out of scope, we'd have to get deinit working for extern records - see issue #15881). Another wrinkle is that while `c_dynamic_array` might be the name to refer to it from Chapel, the name within C would probably need to be more along the lines of `chpl_interop_array` or something (since C doesn't have namespaces or anything).

**Comment by bradcray** [comment]
> I think it'd be better to have one C struct representing a Chapel array for interoperability than two.

I don't view `c_array` or `c_dynamic_array` as representing a Chapel array to C, but rather to represent a C array to Chapel.  I suppose it could be argued that C doesn't actually have dynamic arrays, it just has pointers that can be indexed, but my previous comment was trying to motivate why it seems like it could be an attractive companion to `c_array` for users.

**Comment by BryantLam** [comment]
> I don't view `c_array` or `c_dynamic_array` as representing a Chapel array to C, but rather **to represent a C array to Chapel**.

Both of these arrays are semantically created, owned, and automatically deallocated from Chapel. In a way, they serve as a mechanism for Chapel to present a C array *to* C code. @mppf is arguing in a similar line to your thinking, but the name should be more aptly changed because at least for `c_dynamic_array`, there technically isn't a C-interop reason to have it since it is composed of the baser C pointer and a size like you're suggesting.

**Comment by bradcray** [comment]
That's a good point.  Though it seems if I wrote `extern foo(x: c_[dynamic_]array)` then it could be a case of a C array being passed into Chapel, which is one case I was thinking about.  I was also truly imagining that `c_dynamic_array` would only be a pointer, not a record with a pointer and bundled size.

**Comment by BryantLam** [comment]
Yeah, I think it merits discussion on what is the definition for `c_dynamic_array`. I was arguing for something closer to what Michael envisioned as a `chpl_interop_array` because one of the use cases is to get that Chapel-owned heap array to C code. But as you said, from an interfacing point of view, maybe there's merit to having an `extern foo(x: c_dynamic_array)` be supported, though I think this is where I was pushing in the direction of getting a separate or compatible type to do array views over that memory since the C code owns it.

**Comment by mppf** [comment]
My view on this is that we should have `c_dynamic_array` (or whatever it is called) be a C struct - but that the compiler should allow coercing it into a `c_ptr` (which would just use the ptr part of the struct). That way, it is possible to pass to a function declared as accepting a `c_ptr`. So another way to put it is that I think it's important for `c_dynamic_array` to have the abilities a struct gives (like ability to query the length or know how to free) - but that it should be easy to go between ptr + length and the struct and back again.

> Why not just use c_ptr?

One thing (the main thing?) is that `c_dynamic_array` would manage the memory like other Chapel array types, so that it would not need to be manually deleted when it goes out of scope.

**Comment by BryantLam** [comment]
For an FFI to C, I think having coercions from struct to ptr could be confusing or undesirable. It might make sense to allow it from Chapel code, but even then it becomes circumstantial depending on whether the C function's declaration takes just a ptr or a ptr + length. Maybe it'd be okay if there was some per-extern-function pragma that says "do an interface coercion".

My opinion is that it would be better (for C code anyway) to be explicit and represent the dynamic array in C as a struct with `chpl_*` accessor functions. Anyone that wants the casts should do it on the dynamic array explicitly from the Chapel code or the C code.

**Comment by bradcray** [comment]
On reflection, I think you guys are right that my `extern foo(x: c_dynamic_array)` case wasn't interesting because if the array was originating from C, there's no reason not to just declare such cases as `extern foo(x: c_ptr(t))`.  So I think the main remaining case for this (as I was thinking of it anyway) would be as a convenience over c_ptr + malloc + free within Chapel to get a 1D array of heap-allocated memory, as a sibling to `c_array`.

Once we get into features like the slicing in the OP or the accessor functions in Bryant's latest comments, it becomes less clear to me what the advantages of a `c_dynamic_array` over, say, a Chapel array are (in that it starts to feel less and less C-like).  Put another way, I think of `c_ptr` and `c_array` as being ways of saying "give me something from C within Chapel; whereas once we start talking about functionality beyond what C supports on its types, I'd be more inclined to represent that using Chapel types (potentially with ways of describing themselves out to a C routine similar to the `c_external_array` concept, which I think still needs some review and polishing...).
